src/index/CompressedBlockPrefiltering.cpp

```diff
@@ -0,0 +1,195 @@
+        // and lastTriple we have to include the respective metadata block
+        // (that block is partially relevant).
+        input.begin() + std::distance(valueIdsInput.begin(), secondId + 1) / 2);
+  }
+  relevantBlocks.shrink_to_fit();
+  return relevantBlocks;
+};
+
+// SECTION LOGICAL OPERATIONS
+//______________________________________________________________________________
+template <LogicalOperators Operation>
+std::unique_ptr<PrefilterExpression>
+LogicalExpression<Operation>::logicalComplement() const {
+  using enum LogicalOperators;
+  switch (Operation) {
+    case NOT:
+      // Logically we complement (negate) a NOT here => NOT cancels out.
+      // Therefore, we can simply return the child of the respective NOT
+      // expression after undoing its previous complementation.
+      return child1_->logicalComplement();
+    // Source De-Morgan's laws: De Morgan's laws, Wikipedia.
+    // Reference: https://en.wikipedia.org/wiki/De_Morgan%27s_laws
+    case OR:
+      AD_CONTRACT_CHECK(child2_.has_value());
+      // De Morgan's law: not (A or B) = (not A) and (not B)
+      return std::make_unique<AndExpression>(
+          child1_->logicalComplement(), child2_.value()->logicalComplement());
+    case AND:
+      AD_CONTRACT_CHECK(child2_.has_value());
+      // De Morgan's law: not (A and B) = (not A) or (not B)
+      return std::make_unique<OrExpression>(
+          child1_->logicalComplement(), child2_.value()->logicalComplement());
+    default:
+      AD_FAIL();
+  }
+};
+
+//______________________________________________________________________________
+template <LogicalOperators Operation>
+std::vector<BlockMetadata> LogicalExpression<Operation>::evaluateOr(
+    const std::vector<BlockMetadata>& input, size_t evaluationColumn) const
+    requires(Operation == LogicalOperators::OR) {
+  auto relevantBlocksChild1 = child1_->evaluate(input, evaluationColumn);
+  auto relevantBlocksChild2 =
+      child2_.value()->evaluate(input, evaluationColumn);
+  // The result vector, reserve space for the logical upper bound (num blocks)
+  // |relevantBlocksChild1| + |relevantBlocksChild2| regarding the OR
+  // operation (union of blocks)
+  std::vector<BlockMetadata> unionedRelevantBlocks;
+  unionedRelevantBlocks.reserve(relevantBlocksChild1.size() +
+                                relevantBlocksChild2.size());
+  // Lambda function that implements the projection to retrieve ValueIds from
+  // BlockMetadata values.
+  auto retrieveValueId = [evaluationColumn](const BlockMetadata& block) {
+    return getIdFromColumnIndex(block.lastTriple_, evaluationColumn);
+  };
+  // Given that we have vectors with sorted (BlockMedata) values, we can
+  // use std::ranges::set_union, thus the complexity is O(n + m).
+  std::ranges::set_union(relevantBlocksChild1, relevantBlocksChild2,
+                         std::back_inserter(unionedRelevantBlocks),
+                         std::ranges::less{}, retrieveValueId, retrieveValueId);
```

Comment:
Your comparator has to lexicographically compare the complete blocks (or at least the part containing the relevant column + ALL following columns.
Consider blocks 
(1 3 5 - 1 3 7)           (1 3 29 - 1 3 35)
On the other hand... this can never happen.
But it is important to have good unit tests here.

Reply:
But it would still be more readable to make the retrieve function something like `std::tie(col0Id_, col1Id_, col2Id_)` and then the `std::less` does the lexicographical compare.

Reply:
Note: this requires the assertions that if you compare e.g. col1Ids, that then the col0Id is always the same.

Reply:
And in fact you need a comparison that requires a comparison
` blocka.lastTriple < blockB.firstTriple`  where the `<` is according to the `std::tie`from above.

Reply:
I think there is functionality for this in CompressedRelation.cpp which can be extracted to a separate function `BlockLessThanBlock`

Reply:
maybe this doesn't exist in that form, and I've hallucinated it.